core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1933,63 +1933,78 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
+            val signalStop: F2[Unit] = stopDef.complete(()).void
+            val stop: F2[Either[Throwable, Unit]] = stopDef.get.as(Right(()))
+            def complete(result: Either[Throwable, Unit]): F2[Unit] =
+              bothStates.update {
+                case (None, None)  => (Some(result), None)
+                case (other, None) => (other, Some(result))
+                case _             => sys.error("impossible")
+              }
+            val bothStopped: PartialFunction[(State, State), Either[Throwable, Unit]] = {
+              case (Some(r1), Some(r2)) => CompositeFailure.fromResults(r1, r2)
+            }
+            def run(s: Stream[F2, O2]): F2[Unit] =
+              // `guard` ensures we do not pull another chunk until the previous one has been consumed downstream.
+              Semaphore[F2](1).flatMap { guard =>
+                Stream
+                  .exec(guard.acquire)
+                  .append(s)
+                  .chunks
+                  .evalMap(chk =>
+                    output.send(
+                      Stream.chunk(chk).onFinalize(guard.release)
+                    )
+                  )
+                  .evalTap(_ => guard.acquire)
+                  // Stop when the other upstream has errored or the downstream has completed.
+                  .interruptWhen(stop)
+                  .compile
+                  .drain
+                  .attempt
+                  .flatMap {
+                    case r @ Left(_) =>
+                      // On error, interrupt the other upstream and downstream.
+                      complete(r) >> signalStop
+                    case r @ Right(()) => complete(r)
+                  }
+              }
 
-      def runStream(s: Stream[F2, O2], whenDone: Deferred[F2, Either[Throwable, Unit]]): F2[Unit] =
-        // guarantee we process only single chunk at any given time from any given side.
-        Semaphore(1).flatMap { guard =>
-          val str = watchInterrupted(go(s, guard).stream)
-          str.compile.drain.attempt.flatMap {
-            // signal completion of our side before we will signal interruption,
-            // to make sure our result is always available to others
-            case r @ Left(_)  => whenDone.complete(r) >> signalInterruption
-            case r @ Right(_) => whenDone.complete(r) >> doneAndClose
+            val waitForBoth: F2[Unit] = bothStates.discrete
+              .collect(bothStopped)
+              .head
+              .rethrow
+              .compile
+              .drain
+              .guarantee(output.close.void)
+
+            // There is no need to clean up these fibers. If the downstream is cancelled,
+            // both streams will stop gracefully and the fibers will complete.
+            val setup: F2[Fiber[F2, Throwable, Unit]] =
+              run(this).start >> run(that).start >> waitForBoth.start
+            Stream.bracket(setup)(waitForBoth =>
+              signalStop >> waitForBoth.joinWithUnit
```

Comment:
Avoid name collision.
```suggestion
            Stream.bracket(setup)(wfb => signalStop >> wfb.joinWithUnit)
```